src/Router/UrlGenerator.php

```diff
@@ -1,39 +1,54 @@
+     * @param  bool|null  $secure
+     * @return string
+     */
+    public function to($path, $extra = [], $secure = null)
+    {
+        $url = parent::to($path, $extra, $secure);
+        return static::buildUrl($url);
+    }
+
     /**
      * Build a URL from an array returned from a `parse_url` call.
      *
      * This function serves as a counterpart to the `parse_url` method available in PHP, and a userland implementation
-     * of the `http_build_query` method provided by the PECL HTTP module. This allows a developer to parse a URL to an
+     * of the `http_build_url` method provided by the PECL HTTP module. This allows a developer to parse a URL to an
      * array and make adjustments to the URL parts before combining them into a valid URL reference string.
      *
-     * Based off of the implentation at https://github.com/jakeasmith/http_build_url/blob/master/src/http_build_url.php.
+     * Based off of the implentation at https://github.com/jakeasmith/http_build_url/blob/master/src/http_build_url.php
+     * and https://github.com/ivantcholakov/http_build_url/blob/master/http_build_url.php
+     *
+     * NOTE: Key differences from the PECL implementation include NOT using the current request scheme / host / path to
+     * provide fallbacks for not present values (use Url::to() instead) and supporting providing query parameters as an
+     * array of values rather than just a query string.
      *
-     * @param mixed $url The URL parts, as an array. Must match the structure returned from a `parse_url` call.
-     * @param mixed $replace The URL replacement parts. Allows a developer to replace certain sections of the URL with
+     * @see https://php.uz/manual/en/function.http-build-url.php
+     * @param string|array $url The URL parts, as an array. Must match the structure returned from a `parse_url` call.
+     * @param string|array $replace The URL replacement parts. Allows a developer to replace certain sections of the URL with
      *                       a different value.
      * @param mixed $flags A bitmask of binary or'ed HTTP_URL constants. By default, this is set to HTTP_URL_REPLACE.
      * @param array $newUrl If set, this will be filled with the array parts of the composed URL, similar to the return
      *                      value of `parse_url`.
-     *
      * @return string The generated URL as a string
      */
     public static function buildUrl($url, $replace = [], $flags = HTTP_URL_REPLACE, &$newUrl = []): string
     {
-        if (is_string($url)) {
-            $url = parse_url($url);
-        }
-        if (is_string($replace)) {
-            $replace = parse_url($replace);
-        }
-
         $urlSegments = ['scheme', 'host', 'user', 'pass', 'port', 'path', 'query', 'fragment'];
 
+        // Setup special handling for common schemes
+        // @see https://en.wikipedia.org/wiki/List_of_URI_schemes for future special cases
+        $singleColonSchemes = ['mailto', 'tel', 'sms'];
```

Comment:
I think we might need to try and cater for *any* type of single-colon scheme, as opposed to being explicit. The only one we'd really need to do some sort of handling (or more to the point, denying) is `javascript:`.

Reply:
The reason I'm explicit here is because Laravel is explicit, probably because these are uncommon but still relatively common schemes to use in a web environment. I'm not sure what a generic approach to this would look like without getting really into the weeds, at some point we have to accept that it's impossible to handle every conceivable variation of custom schemes that could be out there and just focus on the ones most likely to affect our users.